### PR TITLE
enabling host code coverage

### PR DESCRIPTION
**Pull request by YvanMokwinski**
using code coverage tools: lcov and gcov.
'--codecoverage' option in install.sh

added an option for release debug mode
'-k | --RelWithDebInfo'

Note: needs to be compiled for a specific GPU architecture, otherwise there is a linkage issue.

example:
./install.sh -c -a gfx906 --codecoverage
or 
./install.sh -kc -a gfx906 --codecoverage

once compilation is done

    cd build/release (or release-debug)
    make coverage GTEST_FILTER=daily*GETRS.__float/*

it takes longer than the normal mode to run the tests

when finished, open index.html in the lcoverage/ directory.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -72,6 +72,9 @@ Options:
                               Building locally with this flag will require docker on your machine. If you are
                               familiar with doxygen, sphinx and documentation tools, you can alternatively
                               use the scripts provided in the docs directory.
+  --codecoverage                   Build with code coverage profiling enabled
+  -k | --relwithdebinfo       Pass this flag to build in release debug mode (equivalent to set CMAKE_BUILD_TYPE=RelWithDebInfo).
+                              (Default build type is Release)
```
Comment (jzuniga-amd) [inline install.sh:77]: Please remove tab after `--codecoverage`
and add a newline between `--codecoverage` and `-k`

**Review by jzuniga-amd — COMMENTED** [review]
When code coverage is enabled you are adding `-g` to compiler options... so, when building in release mode with code coverage, shouldn't you switch the build mode to `relwithdebinfo` instead? 

In other words:
`release` mode = compiler optimizations, no debug info, no code coverage.
`relwithdebinfo` mode  = compiler optimizations + debug info, code coverage optional (enable with `--codecoverage` flag)
`debug` mode = debug info + runtime checks, no optimizations, code coverage optional (enable with `--codecoverage` flag)

File: install.sh (hunk 1 above)
Comment (YvanMokwinski) [reply install.sh:77]: I am also adding a new line before code coverage.

**Comment by YvanMokwinski** [comment]
> When code coverage is enabled you are adding `-g` to compiler options... so, when building in release mode with code coverage, shouldn't you switch the build mode to `relwithdebinfo` instead?
> 
> In other words:
> `release` mode = compiler optimizations, no debug info, no code coverage.
> `relwithdebinfo` mode = compiler optimizations + debug info, code coverage optional (enable with `--codecoverage` flag)
> `debug` mode = debug info + runtime checks, no optimizations, code coverage optional (enable with `--codecoverage` flag)

I understand your point and I have considered it. However, for simplicity reason I kept the possibility of enabling the code coverage for any mode. If you really want to reject the code coverage for release mode  we can do it but to me it has very little benefit, it would need extra checks in the install script and/or the CMakeLists.txt to reject such an option if release mode is enabled.

**Comment by YvanMokwinski** [comment]
Exclusion of the code coverage in Release mode as suggested.

**Review by jzuniga-amd — APPROVED** [review]
LGTM, thanks.
(please wait for Cory's and Troy's reviews, and `squash & merge` after approval)

**Review by tfalders — APPROVED** [review]
LGTM! Should be a handy tool to have.

**Review by cgmb — COMMENTED** [review]
Mostly questions. Though I have a couple requests for minor changes.

File: CMakeLists.txt [hunk 2]
```diff
@@ -129,3 +139,57 @@ endif( )
...
+#
+# ADDITIONAL TARGETS FOR CODE COVERAGE
+#
+# > make coverage_cleanup (clean coverage related files.)
+# > make coverage GTEST_FILTER=<>
+# will run:
+#  > make coverage_analysis GTEST_FILTER=<> (analyze tests)
+#  > make coverage_output (generate html documentation)
+#
+#
```
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:153]: sure

File: CMakeLists.txt [hunk 3]
```diff
@@ -129,3 +139,57 @@ endif( )
...
+# Generate coverage output.
+#
+add_custom_command(TARGET coverage_output
+  COMMAND lcov --directory . --base-directory . --gcov-tool ${CMAKE_BINARY_DIR}/llvm-gcov.sh --capture -o lcoverage/raw_main_coverage.info
+  COMMAND lcov --remove lcoverage/raw_main_coverage.info "'/opt/*'" "'/usr/*'" -o lcoverage/main_coverage.info
+  COMMAND genhtml --ignore-errors source lcoverage/main_coverage.info --output-directory lcoverage
+    )
+
+add_custom_target(coverage DEPENDS coverage_analysis coverage_output)
```
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:187]: okay I will investigate.

File: CMakeLists.txt [hunk 4]
```diff
@@ -129,3 +139,57 @@ endif( )
 if( BUILD_CLIENTS_TESTS OR BUILD_CLIENTS_BENCHMARKS OR BUILD_CLIENTS_SAMPLES )
   add_subdirectory( clients )
 endif( )
+
+
+#
+# ADDITIONAL TARGETS FOR CODE COVERAGE
+#
+# > make coverage_cleanup (clean coverage related files.)
+# > make coverage GTEST_FILTER=<>
```
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:148]: it's a variable I'm passing to rocsolver-test.

File: CMakeLists.txt [hunk 5]
```diff
@@ -129,3 +139,57 @@ endif( )
...
+  COMMAND genhtml --ignore-errors source lcoverage/main_coverage.info --output-directory lcoverage
+    )
+
+add_custom_target(coverage DEPENDS coverage_analysis coverage_output)
+
+#
+# Coverage cleanup
+#
+add_custom_target(coverage_cleanup
+    COMMAND rm -f `find ${CMAKE_BINARY_DIR} -name *.gcda`
```
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:193]: I assume there is many ways to do that, I chose one of them.

[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:193]: I don't know if your command works.

File: CMakeLists.txt [hunk 6]
```diff
@@ -91,6 +91,16 @@ option( BUILD_CLIENTS_TESTS "Build rocSOLVER unit tests" OFF )
 option( BUILD_CLIENTS_BENCHMARKS "Build rocSOLVER benchmarks" OFF )
 option( BUILD_CLIENTS_SAMPLES "Build rocSOLVER samples" OFF )
 
+# FOR OPTIONAL CODE COVERAGE
+option(BUILD_CODE_COVERAGE "Build rocSOLVER with code coverage enabled" OFF)
+if(BUILD_CODE_COVERAGE)
+  if( CMAKE_BUILD_TYPE STREQUAL "Release" )
+    message(FATAL_ERROR "error: code coverage is not supported in Release mode.")
+  endif()
```
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:99]: I would keep it.
some people loves to skip the install script.

File: CMakeLists.txt [hunk 7]
```diff
@@ -129,3 +139,57 @@ endif( )
...
+
+#
+# Prepare coverage output
+#
+add_custom_target(coverage_output
+    COMMAND mkdir -p lcoverage
+    COMMAND echo "\\#!/bin/bash" > llvm-gcov.sh
+    COMMAND echo "\\# THIS FILE HAS BEEN GENERATED" >> llvm-gcov.sh
+    COMMAND printf "exec /opt/rocm/llvm/bin/llvm-cov gcov $$\\@" >> llvm-gcov.sh
+    COMMAND chmod +x llvm-gcov.sh
```
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:175]: yes

File: CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:196]: which weird edge cases?

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:178]: added a comment to make it clear.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt]: done

File: CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:190]: I see in your log file you didn't run any test, that's why it failed (as far as I know).
The purpose of GTEST_FILTER is to be able to select some tests to run, here basically you're running 'rocsolver-test --gtest_filter=' because the GTEST_FILTER is variable empty. 

This variable It's convenient when you're working on a given routine and you don't want to run all the tests.

Moreover, for instance in rocSPARSE I targeted pre_checkin tests only for CI code coverage, so we are using the --gtest_filter.
I didn't really see the benefit of skipping --gtest_filter= if the GTEST_FILTER variable is empty.

[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:190]: Actually, no you're right something is not setup properly.

File: CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt]: anyway -delete is more elegant, I'm still interested to know why this is not equivalent.

**Comment by YvanMokwinski** [comment]
@cgmb Good review ! going to forward these improvements to rocBLAS and rocSPARSE

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (YvanMokwinski) [reply CMakeLists.txt:147]: The purpose of GTEST_FILTER is to be able to select some tests to run.
This variable It's convenient when you're working on a given routine and you don't want to run all the tests.
If you don't specify it, it won't run any test.

File: CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (cgmb) [reply CMakeLists.txt]: > which weird edge cases?

They're unlikely to be a problem in this case, but:
- Names that begin with `-`
- Names with spaces in them

There's [a section in the GNU find manual](https://www.gnu.org/software/findutils/manual/html_node/find_html/Deleting-Files.html) that discusses various ways of doing this.

**Review by cgmb — APPROVED** [review]
Please wait for the PreCheckin tests to pass, then go ahead with a squash-and-merge. The build is currently broken, but should be fixed by #254.

**Comment by cgmb** [comment]
The CI was a bit of a mess most of the day, so I ran what I could and merged.

Thanks, Yvan! It's a great tool to have available.
